internal/controller/istio_extension_reconciler_test.go

```diff
@@ -0,0 +1,547 @@
+
+				// Verify the rate limit action was properly merged
+				if mergedRateLimitAction != nil {
+					assert.Equal(t, 2, len(mergedRateLimitAction.ConditionalData), "merged rate limit action should have 2 conditional data items")
+				}
+			},
+		},
+		{
+			name: "multiple auth actions with different scopes - never merged",
+			actions: []wasm.Action{
+				{
+					ServiceName: wasm.AuthServiceName,
+					Scope:       "global",
+					ConditionalData: []wasm.ConditionalData{
+						{
+							Data: []wasm.DataType{
+								{
+									Value: &wasm.Static{
+										Static: wasm.StaticSpec{
+											Key:   "auth.identity",
+											Value: "global_user",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				{
+					ServiceName: wasm.AuthServiceName,
+					Scope:       "namespace",
+					ConditionalData: []wasm.ConditionalData{
+						{
+							Data: []wasm.DataType{
+								{
+									Value: &wasm.Static{
+										Static: wasm.StaticSpec{
+											Key:   "auth.identity",
+											Value: "namespace_user",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedLen: 2,
+			description: "should never merge auth actions regardless of scope differences",
+		},
+		{
+			name: "complex mixed scenario with multiple service types",
+			actions: []wasm.Action{
+				{
+					ServiceName: "ratelimit-service",
+					Scope:       "global",
+					ConditionalData: []wasm.ConditionalData{
+						{
+							Data: []wasm.DataType{
+								{
+									Value: &wasm.Static{
+										Static: wasm.StaticSpec{
+											Key:   "ratelimit.limit_key",
+											Value: "user_id",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				{
+					ServiceName: wasm.AuthServiceName,
+					Scope:       "global",
+					ConditionalData: []wasm.ConditionalData{
+						{
+							Data: []wasm.DataType{
+								{
+									Value: &wasm.Static{
+										Static: wasm.StaticSpec{
+											Key:   "auth.identity",
+											Value: "jwt_token",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				{
+					ServiceName: "ratelimit-service",
+					Scope:       "global",
+					ConditionalData: []wasm.ConditionalData{
+						{
+							Data: []wasm.DataType{
+								{
+									Value: &wasm.Expression{
+										ExpressionItem: wasm.ExpressionItem{
+											Key:   "tokenratelimit.limit_key",
+											Value: "api_key",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				{
+					ServiceName: wasm.AuthServiceName,
+					Scope:       "global",
+					ConditionalData: []wasm.ConditionalData{
+						{
+							Data: []wasm.DataType{
+								{
+									Value: &wasm.Static{
+										Static: wasm.StaticSpec{
+											Key:   "auth.identity",
+											Value: "user_id",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedLen: 3,
```

Comment:
the expected length is 4. Only sub-sequent actions can be merged. Due to data dependency rules, you never know if one rate limit action needs a pre-auth action.

Reply:
gotcha, changed the logic to only merge subsequent actions with same service names